rfd/0197-prometheus-metrics.md

```diff
@@ -0,0 +1,237 @@
+    }
+    
+    errs := trace.NewAggregate(
+        reg.Register(m.currentFoo),
+        reg.Register(m.barCounter),
+    )
+    
+    if errs != nil {
+        return trace.Wrap(err, "registering metrics")
+    }
+    
+    return &m
+}
+
+```
+
+#### Don't
+
+- Store metrics in a package-scoped variable
+
+```golang
+var (
+    currentFoo = prometheus.NewGauge(prometheus.GaugeOpts{
+            Namespace: teleport.MetricNamespace,
+            Subsystem: metricsSubsystem,
+            Name: "foo_current",
+            Help: "Measures the number of foos.",
+        })
+    // ...
+)
+```
+
+### Naming metrics
+
+#### Do
+
+- Use `teleport.MetricsNamespace` as the namespace.
+- Use a subsystem name unique to your component.
+- Follow [the prometheus metrics naming guidelines](https://prometheus.io/docs/practices/naming/),
+  especially always specify the unit and use a suffix to clarify the metric type (`_total`, `_info`).
+
+```golang
+type metrics struct {
+    currentFoo prometheus.Gauge
+    barCounter *prometheus.CounterVec
+}
+
+const metricsSubsystem = "my_service"
+
+func newMetrics(reg prometheus.Registerer) (*metrics, error) {
+    m := metrics{
+        currentFoo: prometheus.NewGauge(prometheus.GaugeOpts{
+            Namespace: teleport.MetricNamespace,
+            Subsystem: metricsSubsystem,
+            Name: "foo_timestamp_seconds",
+            Help: "Represents the foo time, in seconds.",
+        }),
+        barCounter: prometheus.NewCounter(prometheus.GaugeOpts{
+            Namespace: teleport.MetricNamespace,
+            Subsystem: metricsSubsystem,
+            Name: "bar_total",
+            Help: "Number of times bar happened.",
+        }),
+    }
+    // ...
+}
+```
+
+#### Don't
+
+- Manually namespace metrics
+- Create non-namespaced metrics
+
+```golang
+type metrics struct {
+    currentFoo prometheus.Gauge
+    barCounter *prometheus.CounterVec
+}
+
+func newMetrics(reg prometheus.Registerer) (*metrics, error) {
+    m := metrics{
+        currentFoo: prometheus.NewGauge(prometheus.GaugeOpts{
+            Name: "teleport_foo_timestamp_seconds",
+            Help: "Represents the foo time, in seconds.",
+        }),
+        barCounter: prometheus.NewCounter(prometheus.GaugeOpts{
+            Name: "bar_total",
+            Help: "Number of times bar happened.",
+        }),
+    }
+    // ...
+}
+```
+
+### Metric registration
+
+#### Do
+
+- Use `reg.Register()` to register the metric.
+- Aggregate errors and fail early if you can't register metrics.
+
+#### Don't
+
+- Use the package helpers `prometheus.Register()` or `prometheus.MustRegister`.
+- Use `reg.MustRegister` as it panics in case of conflict and doesn't clearly indicate which metric descriptor conflicted.
+
+### Labeling
+
+#### Do
+
+- Ensure no secret information is present in the labels.
+- Avoid high cardinality labels
+
+#### Don't
+
+- Put user input directly in the labels
+- Create large (1k+) metric combinations
+
```

Comment:
Should we talk about metric conflicts? Global vs local, local vs local, in what situations it errors vs it gets "summed", etc?

What happens if a gauge conflicts?

Anything else we need to know?